extension/src/openvic-extension/singletons/CursorSingleton.cpp

```diff
@@ -0,0 +1,602 @@
+
+	//ani header variables
+	int32_t numFrames = 1;
+	int32_t numSteps = 1;
+	Vector2i dimensions = Vector2i(1,1);
+	int32_t bitCount = 1;
+	int32_t numPlanes = 1; //???
+	int32_t displayRate = 1; //how long each frame should last
+	int32_t flags = 0;
+	bool iconFlag = false;
+	bool sequenceFlag = false;
+
+
+	while(file->get_position() < riff_size){
+		String id = read_riff_str(file);
+		int32_t size = file->get_32();
+		if(id == "LIST"){
+			String list_type = read_riff_str(file);
+		}
+		else if(id == "anih"){
+			//hack for some files, there's likely a better way
+			if(size == 36){ 
+				int32_t headerSize = file->get_32();
+			}
+			numFrames = file->get_32();
+			numSteps = file->get_32();
+			dimensions = Vector2i(file->get_32(),file->get_32());
+			bitCount = file->get_32();
+			numPlanes = file->get_32();
+			displayRate = file->get_32();
+			flags = file->get_32();
+			iconFlag = flags & 0x1;
+			sequenceFlag = flags & 0x2;
+		}
+		else if(id == "icon"){
+
+			int32_t file_access_offset = file->get_position();
+
+			image_hotspot_pair_asset_t pair = _load_pair(file);
+			//basically pushback an array
+
+			//only store the resolutions from one frame
+			if(resolutions.is_empty()){
+				for(int32_t i=0;i<pair.images.size();i++){
+					
+					PackedVector2Array hotspots;
+					TypedArray<ImageTexture> images;
+					images.push_back(pair.images[i]);
+					hotspots.push_back(pair.hotspots[i]);
+					frames_by_resolution.push_back(images);
+					hotspots_by_resolution.push_back(hotspots);
+
+					resolutions.push_back(Vector2i(pair.images[i]->get_width(),pair.images[i]->get_height()));
+				}
+
+			}
+			else {
+				for(int32_t i=0;i<pair.images.size();i++){
+					frames_by_resolution[i].push_back(pair.images[i]);
+					hotspots_by_resolution[i].push_back(pair.hotspots[i]);
+				}
+			}
+
+			//cursor could have been anywhere in the file, come back to a known position
+			file->seek(file_access_offset + size);
+
+		}
+		else if(id == "seq "){
+			for(int32_t i=0;i<numSteps;i++){
+				sequence.push_back(file->get_32());
+			}
+		}
+		else if(id == "rate"){
+			for(int32_t i=0;i<numSteps;i++){
+				displayRates.push_back(file->get_32()/60.0);
+			}
+		}
+		else {
+			//Various junk (JUNK, metadata we don't care about, ...)
+			file->get_buffer(size);
+		}
+		//align to even bytes
+		if(file->get_position() % 2 != 0){
+			file->get_8();
+		}
+	}
+
+	//not all ani files have the sequence and rate chunks, if not, fill out these properties
+	//manually
+	if(sequence.size() == 0){
+		for(int32_t i=0; i<numSteps;i++){
+			sequence.push_back(i);
+		}
+	}
+	if(displayRates.size() == 0){
+		for(int32_t i=0; i<numSteps;i++){
+			displayRates.push_back(displayRate/60.0);
+		}
+	}
+
+	cursor_asset_t cursor = {
+		hotspots_by_resolution,
+		frames_by_resolution,
+		resolutions,
+		static_cast<int>(sequence.size())
+	};
+
+	cursor.displayRates = displayRates;
+	cursor.sequence = sequence;
+	cursors.emplace(std::move(name),cursor);
```

Comment:
```suggestion
	cursors.emplace(
		name,
		cursor_asset_t {
			std::move(hotspots_by_resolution),
			std::move(frames_by_resolution),
			resolutions,
			static_cast<int32_t>(sequence.size()),
			displayRates,
			sequence
		}
	);
```
- We can `std::move` the `hotspots_by_resolution` and `frames_by_resolution` vectors into the `cursor_asset_t` rather than copying them.
- We can include `displayRates` and `sequence` in the intialiser list rather than setting then manually after construction.
- We could `std::move` `cursor` into `cursors` to avoid copying it (particularly its vector member variables), although it would be simplest to just construct it in place.
- We can remove the `std::move` on `name` as it has no effect given `name` is a `const&` variable (forces a copy as there's no mutable instance to steal allocated memory from) and anyway with `StringName`s there's nothing saved by moving.